src/solvers/dg_unstructured_quad/containers_2d.jl

```diff
@@ -0,0 +1,322 @@
+
+  return elements
+end
+
+
+# initialize all the values in the container on a straight sided element
+function init_element!(elements, element_id, polydeg, nodes, corners)
+
+  calc_node_coordinates!(elements.node_coordinates, element_id, nodes, polydeg, corners)
+
+  calc_metric_terms!(elements.X_xi, elements.X_eta, elements.Y_xi, elements.Y_eta, element_id,
+                     nodes, polydeg, corners)
+
+  calc_inverse_jacobian!(elements.inverse_jacobian, element_id, elements.X_xi, elements.X_eta,
+                         elements.Y_xi, elements.Y_eta)
+
+  calc_normals_scaling_and_tangents!(elements.normals, elements.scaling, elements.tangents,
+                                     element_id, nodes, polydeg, corners)
+
+  return elements
+end
+
+
+# generic container for the interior interfaces of an unstructured mesh
+struct UnstructuredInterfaceContainer2D{RealT<:Real, NVARS, POLYDEG}
+  u                ::Array{RealT, 4} # [primary/secondary, variables, i, interfaces]
+  start_index      ::Vector{Int64}   # [interfaces]
+  inc_index        ::Vector{Int64}   # [interfaces]
+  element_ids      ::Array{Int64, 2} # [primary/secondary, interfaces]
+  element_side_ids ::Array{Int64, 2} # [primary/secondary, interfaces]
+end
+
+
+# construct an empty curved interface container to be filled later with neighbour information in the
+# unstructured mesh constructor
+function UnstructuredInterfaceContainer2D{RealT, NVARS, POLYDEG}(capacity::Int64) where {RealT<:Real, NVARS, POLYDEG}
+
+  n_nodes = POLYDEG + 1
+  nan = convert(RealT, NaN)
+
+  u                = fill(nan, (2, NVARS, n_nodes, capacity))
+  start_index      = fill(typemin(Int64), capacity)
+  inc_index        = fill(typemin(Int64), capacity)
+  element_ids      = fill(typemin(Int64), (2, capacity))
+  element_side_ids = fill(typemin(Int64), (2, capacity))
+
+  return UnstructuredInterfaceContainer2D{RealT, NVARS, POLYDEG}(u, start_index, inc_index,
+                                                                 element_ids, element_side_ids)
+end
+
+
+@inline ninterfaces(interfaces::UnstructuredInterfaceContainer2D) = length(interfaces.start_index)
+
+@inline eachinterface(interfaces::UnstructuredInterfaceContainer2D) = Base.OneTo(ninterfaces(interfaces))
+
+
+function init_interfaces(RealT, mesh, nvars, poly_deg)
+
+  # get the number of internal interfaces in the mesh
+  if isperiodic(mesh)
+    interior_interfaces = mesh.n_interfaces
+  else
+    interior_interfaces = mesh.n_interfaces - mesh.n_boundary
+  end
+
+  interfaces = UnstructuredInterfaceContainer2D{RealT, nvars, poly_deg}(interior_interfaces)
+
+  # extract and save the appropriate neighbour information from the mesh skeleton
+  init_interfaces!(interfaces, mesh.neighbour_information, mesh.boundary_names, poly_deg,
+                   mesh.n_elements, Val(isperiodic(mesh)))
+
+  return interfaces
+end
+
+
+function init_interfaces!(interfaces, edge_information, boundary_names, polydeg, n_elements,
```

Comment:
I really don't understand what is going on in this method. I think it would be good to add some comments (some general, introductory ones as well as some inline code description), since I feel like a lot of information about how elements are coupled and how the different coordinate systems work is encoded here. Maybe we can walk through this code at some point such that I also get an idea how this all works ;-)